docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,356 @@
+# Supporting dependencies between build artifacts
+
+## Background
+Refer [this](https://tinyurl.com/skaffold-modules) document which presents the design goals for introducing the concept of _modules_ in Skaffold. A prerequisite to being able to define modules and supporting cross module dependency is to first consider the latest version of the skaffold config (_currently `v2beta8`_) as an implicit module and allowing dependencies between artifacts defined in it. The current document aims to capture the major code changes necessary for achieving this.
+
+>*Note: Omitted some details for brevity, assuming reader's familiarity with the skaffold [codebase](https://github.com/GoogleContainerTools/skaffold).*
+
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
+```go
+type ArtifactDependency struct {
+	ImageName string `yaml:"image" yamltags:"required"`
+	Alias string `yaml:"alias,omitempty"`
+}
+```
+This allows us to define a build stanza like below where image `leeroy-app` requires image `simple-go-app`:
+```yaml
+build:
+ artifacts:
+   - image: simple-go-app
+   - image: leeroy-app
+     requires:
+       - image: simple-go-app
+         alias: BASE
+```
+Alias is a token that will be replaced with the image reference in the builder definition files. If no value is provided for `alias` then it defaults to the value of `image`.
```

Comment:
Some builders may not support this as they may have restrictions on alias names.  For example, the docker and custom builders require aliases to be valid environment variable names and so do not allow `/` or `.`.

Reply:
if `alias` is missing and `image` name is not in the correct format, then it'll fail the validation immediately and the user can add an explicit alias then.